File: JANUARY/14-01-2026/3454.dart
class Solution {
  double separateSquares(List<List<int>> squares) {
    final n = squares.length;
    final xs = <int>[];
    final events = <List<dynamic>>[];

    for (var s in squares) {
      final x = s[0];
      final y = s[1];
      final l = s[2];
      xs.add(x);
      xs.add(x + l);
      events.add([y, x, x + l, 1]);
      events.add([y + l, x, x + l, -1]);
    }

    xs.sort();
    final uniq = <int>[];
    for (var v in xs) {
      if (uniq.isEmpty || uniq.last != v) uniq.add(v);
    }
    final m = uniq.length;
    final idx = <int, int>{};
    for (var i = 0; i < m; i++) idx[uniq[i]] = i;

    final cnt = List<int>.filled(4 * m, 0);
    final seglen = List<int>.filled(4 * m, 0);

    void upd(int i, int l, int r, int ql, int qr, int v) {
      if (qr <= l || r <= ql) return;
      if (ql <= l && r <= qr) {
        cnt[i] += v;
      } else {
        final mid = (l + r) >> 1;
        upd(i << 1, l, mid, ql, qr, v);
        upd(i << 1 | 1, mid, r, ql, qr, v);
      }
      if (cnt[i] > 0) {
        seglen[i] = uniq[r] - uniq[l];
      } else if (l + 1 == r) {
        seglen[i] = 0;
      } else {
        seglen[i] = seglen[i << 1] + seglen[i << 1 | 1];
      }
    }

    events.sort((a, b) => a[0].compareTo(b[0]));

    var prev = events[0][0] as int;
    double total = 0.0;
    var i = 0;

    while (i < events.length) {
      final y = events[i][0] as int;
      final dy = y - prev;
      if (dy > 0 && seglen[1] > 0) total += seglen[1] * dy.toDouble();

      while (i < events.length && events[i][0] == y) {
        final e = events[i];
        upd(1, 0, m, idx[e[1]]!, idx[e[2]]!, e[3] as int);
        i++;
      }
      prev = y;
    }

    final half = total / 2.0;
    for (var j = 0; j < cnt.length; j++) cnt[j] = 0;
    for (var j = 0; j < seglen.length; j++) seglen[j] = 0;

    prev = events[0][0] as int;
    double cur = 0.0;
    i = 0;

    while (i < events.length) {
      final y = events[i][0] as int;
      final dy = y - prev;
      if (dy > 0 && seglen[1] > 0) {
        final area = seglen[1] * dy.toDouble();
        if (cur + area >= half) return prev + (half - cur) / seglen[1];
        cur += area;
      }

      while (i < events.length && events[i][0] == y) {
        final e = events[i];
        upd(1, 0, m, idx[e[1]]!, idx[e[2]]!, e[3] as int);
        i++;
      }
      prev = y;
    }

    return prev.toDouble();
  }
}

File: JANUARY/14-01-2026/3454.go
func separateSquares(squares [][]int) float64 {
	n := len(squares)

	xs := make([]int64, 0, 2*n)
	type Event struct {
		y, x1, x2 int64
		t         int64
	}
	events := make([]Event, 0, 2*n)

	for _, s := range squares {
		x, y, l := int64(s[0]), int64(s[1]), int64(s[2])
		xs = append(xs, x, x+l)
		events = append(events,
			Event{y, x, x + l, 1},
			Event{y + l, x, x + l, -1},
		)
	}

	sort.Slice(xs, func(i, j int) bool { return xs[i] < xs[j] })
	uniq := xs[:0]
	for _, v := range xs {
		if len(uniq) == 0 || uniq[len(uniq)-1] != v {
			uniq = append(uniq, v)
		}
	}
	xs = uniq
	m := len(xs)

	idx := make(map[int64]int, m)
	for i, v := range xs {
		idx[v] = i
	}

	cnt := make([]int64, 4*m)
	seglen := make([]int64, 4*m)

	var upd func(i, l, r, ql, qr int, v int64)
	upd = func(i, l, r, ql, qr int, v int64) {
		if qr <= l || r <= ql {
			return
		}
		if ql <= l && r <= qr {
			cnt[i] += v
		} else {
			mid := (l + r) >> 1
			upd(i<<1, l, mid, ql, qr, v)
			upd(i<<1|1, mid, r, ql, qr, v)
		}
		if cnt[i] > 0 {
			seglen[i] = xs[r] - xs[l]
		} else if l+1 == r {
			seglen[i] = 0
		} else {
			seglen[i] = seglen[i<<1] + seglen[i<<1|1]
		}
	}

	sort.Slice(events, func(i, j int) bool { return events[i].y < events[j].y })

	prev := events[0].y
	total := float64(0)
	i := 0

	for i < len(events) {
		y := events[i].y
		dy := y - prev
		if dy > 0 && seglen[1] > 0 {
			total += float64(seglen[1]) * float64(dy)
		}
		for i < len(events) && events[i].y == y {
			upd(1, 0, m, idx[events[i].x1], idx[events[i].x2], events[i].t)
			i++
		}
		prev = y
	}

	half := total / 2.0
	cnt = make([]int64, 4*m)
	seglen = make([]int64, 4*m)

	prev = events[0].y
	cur := float64(0)
	i = 0

	for i < len(events) {
		y := events[i].y
		dy := y - prev
		if dy > 0 && seglen[1] > 0 {
			area := float64(seglen[1]) * float64(dy)
			if cur+area >= half {
				return float64(prev) + (half-cur)/float64(seglen[1])
			}
			cur += area
		}
		for i < len(events) && events[i].y == y {
			upd(1, 0, m, idx[events[i].x1], idx[events[i].x2], events[i].t)
			i++
		}
		prev = y
	}

	return float64(prev)
}